File: lib/bounded_authority_protocol/v1/bounds.ex
defmodule BoundedAuthorityProtocol.V1.Bounds do
  @moduledoc """
  Resource ceilings for the v1 wire profile.

  Callers may tighten resource ceilings. Cryptographic widths are immutable.
  """

  @enforce_keys [
    :compact_bytes,
    :encoded_segment_bytes,
    :decoded_segment_bytes,
    :json_bytes,
    :depth,
    :object_members,
    :array_items,
    :total_nodes,
    :string_bytes,
    :key_bytes,
    :number_lexeme_bytes,
    :integer_magnitude,
    :float_magnitude,
    :kid_bytes,
    :jcs_bytes,
    :uri_bytes,
    :identifier_bytes,
    :nonce_bytes,
    :method_bytes,
    :operation_bytes,
    :audiences,
    :operations,
    :selectors,
    :path_segments,
    :one_of_values,
    :public_key_bytes,
    :signature_bytes,
    :digest_bytes,
    :clock_skew,
    :proof_max_age,
    :chain_row_bytes,
    :chain_rows,
    :anchor_bytes,
    :archive_header_bytes,
    :archive_chunks,
    :archive_bytes,
    :object_version_bytes,
    :key_transitions
  ]
  defstruct @enforce_keys

  @type t :: %__MODULE__{
          compact_bytes: pos_integer(),
          encoded_segment_bytes: pos_integer(),
          decoded_segment_bytes: pos_integer(),
          json_bytes: pos_integer(),
          depth: pos_integer(),
          object_members: pos_integer(),
          array_items: pos_integer(),
          total_nodes: pos_integer(),
          string_bytes: pos_integer(),
          key_bytes: pos_integer(),
          number_lexeme_bytes: pos_integer(),
          integer_magnitude: pos_integer(),
          float_magnitude: pos_integer(),
          kid_bytes: pos_integer(),
          jcs_bytes: pos_integer(),
          uri_bytes: pos_integer(),
          identifier_bytes: pos_integer(),
          nonce_bytes: pos_integer(),
          method_bytes: pos_integer(),
          operation_bytes: pos_integer(),
          audiences: pos_integer(),
          operations: pos_integer(),
          selectors: pos_integer(),
          path_segments: pos_integer(),
          one_of_values: pos_integer(),
          public_key_bytes: pos_integer(),
          signature_bytes: pos_integer(),
          digest_bytes: pos_integer(),
          clock_skew: pos_integer(),
          proof_max_age: pos_integer(),
          chain_row_bytes: pos_integer(),
          chain_rows: pos_integer(),
          anchor_bytes: pos_integer(),
          archive_header_bytes: pos_integer(),
          archive_chunks: pos_integer(),
          archive_bytes: pos_integer(),
          object_version_bytes: pos_integer(),
          key_transitions: pos_integer()
        }

  @maximum %{
    compact_bytes: 65_536,
    encoded_segment_bytes: 32_768,
    decoded_segment_bytes: 24_576,
    json_bytes: 65_536,
    depth: 32,
    object_members: 64,
    array_items: 256,
    total_nodes: 4_096,
    string_bytes: 8_192,
    key_bytes: 128,
    number_lexeme_bytes: 64,
    integer_magnitude: 9_007_199_254_740_991,
    float_magnitude: 9_007_199_254_740_991,
    kid_bytes: 128,
    jcs_bytes: 65_536,
    uri_bytes: 8_192,
    identifier_bytes: 512,
    nonce_bytes: 512,
    method_bytes: 32,
    operation_bytes: 128,
    audiences: 64,
    operations: 64,
    selectors: 64,
    path_segments: 32,
    one_of_values: 256,
    public_key_bytes: 32,
    signature_bytes: 64,
    digest_bytes: 32,
    clock_skew: 60,
    proof_max_age: 300,
    chain_row_bytes: 4_096,
    chain_rows: 65_536,
    anchor_bytes: 8_192,
    archive_header_bytes: 8_192,
    archive_chunks: 65_796,
    archive_bytes: 270_820_384,
    object_version_bytes: 512,
    key_transitions: 256
  }
  @fixed_width_keys [:digest_bytes, :public_key_bytes, :signature_bytes]

  @doc "Returns the immutable v1 profile maxima."
  @spec maximum() :: t()
  def maximum, do: struct!(__MODULE__, @maximum)

  @doc "Builds bounds from a map of caller-supplied tightening overrides."
  @spec new(map()) :: {:ok, t()} | {:error, :invalid}
  def new(overrides \\ %{})

  def new(overrides) when is_map(overrides) do
    if valid_keys_and_values?(Map.to_list(overrides)) do
      {:ok, struct!(__MODULE__, Map.merge(@maximum, overrides))}
    else
      {:error, :invalid}
    end
  end

  def new(_overrides), do: {:error, :invalid}

  @spec coerce(t() | map()) :: {:ok, t()} | {:error, :invalid}
  def coerce(%__MODULE__{} = bounds) do
    bounds
    |> Map.from_struct()
    |> new()
  end

  def coerce(overrides), do: new(overrides)

  defp valid_keys_and_values?([]), do: true

  defp valid_keys_and_values?([{key, value} | rest]) do
    case Map.fetch(@maximum, key) do
      {:ok, maximum} when is_integer(maximum) ->
        valid_override?(key, value, maximum) and valid_keys_and_values?(rest)

      :error ->
        false
    end
  end

  defp valid_integer_override?(value, maximum),
    do: is_integer(value) and value > 0 and value <= maximum

  defp valid_override?(key, value, maximum) when key in @fixed_width_keys,
    do: value == maximum

  defp valid_override?(_key, value, maximum), do: valid_integer_override?(value, maximum)
end

File: sdks/python/src/bounded_authority_verifier/bounds.py
"""The fixed v1 profile maxima — the 38-row Hard maxima table (protocol-v1.md:352-390).

Pinned by the corpus ``bounds.new`` cases so any mistyped constant fails. Fixed-width keys
(REQ1-BOUNDS-fixed-widths, L395): ``digest_bytes``, ``public_key_bytes``, ``signature_bytes`` are the
immutable cryptographic constants of suite BAP1-Ed25519-SHA256 — they MUST equal the maximum exactly
(widening is forbidden); all others may be tightened to a positive integer at most the maximum.
"""

from __future__ import annotations

from collections.abc import Mapping
from types import MappingProxyType

# The 38-row Hard maxima table (protocol-v1.md:352-390). Derived from the spec, NOT from lib/*.ex.
MAXIMA: Mapping[str, int] = {
    "compact_bytes": 65536,
    "encoded_segment_bytes": 32768,
    "decoded_segment_bytes": 24576,
    "json_bytes": 65536,
    "depth": 32,
    "object_members": 64,
    "array_items": 256,
    "total_nodes": 4096,
    "string_bytes": 8192,
    "key_bytes": 128,
    "number_lexeme_bytes": 64,
    "integer_magnitude": 9007199254740991,
    "float_magnitude": 9007199254740991,
    "kid_bytes": 128,
    "jcs_bytes": 65536,
    "uri_bytes": 8192,
    "identifier_bytes": 512,
    "nonce_bytes": 512,
    "method_bytes": 32,
    "operation_bytes": 128,
    "audiences": 64,
    "operations": 64,
    "selectors": 64,
    "path_segments": 32,
    "one_of_values": 256,
    "public_key_bytes": 32,
    "signature_bytes": 64,
    "digest_bytes": 32,
    "clock_skew": 60,
    "proof_max_age": 300,
    "chain_row_bytes": 4096,
    "chain_rows": 65536,
    "anchor_bytes": 8192,
    "archive_header_bytes": 8192,
    "archive_chunks": 65796,
    "archive_bytes": 270820384,
    "object_version_bytes": 512,
    "key_transitions": 256,
}

MaximaKey = str
"""A key into the MAXIMA table (a string, since Python's type system treats it lexically)."""

# The fixed-width keys (REQ1-BOUNDS-fixed-widths, L395) — cannot be tightened or widened.
FIXED_WIDTH_KEYS: frozenset[str] = frozenset({"digest_bytes", "public_key_bytes", "signature_bytes"})


class Bounds:
    """The immutable maximum + an optional per-key tightening override.

    ``maximum`` is always the full MAXIMA table; ``overrides`` carries caller tightenings as a
    READ-ONLY view (``MappingProxyType``) so a caller cannot mutate a shared ``Bounds`` to widen the
    profile maximum globally. The profile maxima is the ceiling for every key; an override may only
    lower a non-fixed key (or set a fixed-width key to exactly its maximum, a no-op).
    """

    __slots__ = ("_overrides",)

    def __init__(self, overrides: Mapping[str, int] | None = None) -> None:
        # Wrap in a read-only MappingProxyType so the overrides cannot be mutated post-construction
        # (a caller doing ``bounds_maximum().overrides["depth"] = 999`` raises TypeError, not widens).
        self._overrides: Mapping[str, int] = MappingProxyType(dict(overrides)) if overrides else MappingProxyType({})

    @property
    def overrides(self) -> Mapping[str, int]:
        """The read-only override map (a MappingProxyType — mutation raises TypeError)."""
        return self._overrides

    def resolve(self, key: str) -> int:
        """Resolve a bound: the override if present, else the maximum."""
        return self._overrides.get(key, MAXIMA[key])


MAXIMUM_BOUNDS = Bounds()
"""The immutable maximum — ``Bounds.maximum()`` (REQ1-BOUNDS-tighten-only)."""


def bounds_maximum() -> Bounds:
    """Return the profile maxima (REQ1-BOUNDS-tighten-only)."""
    return MAXIMUM_BOUNDS


def bounds_resolve(b: Bounds, key: str) -> int:
    """Module-level resolve helper: the override if present, else the maximum."""
    return b.resolve(key)


def bounds_new(tightening: Mapping[str, int] | None = None) -> Bounds:
    """Construct a Bounds from a tightening map.

    REQ1-BOUNDS-reject-list (L397): unknown, non-integer, zero, negative, widening, or fixed-width-
    changing limits are invalid. Fixed-width keys (L395) cannot be tightened or widened, but setting
    to the exact maximum is an identity no-op (valid). Any other value rejects.
    """
    # Local import to avoid a cycle: bounds is imported by error-adjacent primitives, and fail lives
    # in error. (fail does not import bounds, so there is no actual cycle, but the locality keeps the
    # dependency direction explicit.)
    from .error import fail

    if tightening is None:
        return MAXIMUM_BOUNDS
    overrides: dict[str, int] = {}
    for key, value in tightening.items():
        if key not in MAXIMA:
            fail(f"bounds.new: unknown limit {key}")
        if not isinstance(value, int) or isinstance(value, bool):
            fail(f"bounds.new: non-integer limit {key}")
        if key in FIXED_WIDTH_KEYS:
            # REQ1-BOUNDS-fixed-widths: fixed-width keys cannot be tightened or widened, but setting
            # to the exact maximum is an identity no-op (valid). Any other value rejects.
            if value != MAXIMA[key]:
                fail(f"bounds.new: fixed-width key {key} must equal maximum")
        else:
            if value <= 0:
                fail(f"bounds.new: non-positive limit {key}")
            if value > MAXIMA[key]:
                fail(f"bounds.new: widening limit {key}")
        overrides[key] = value
    return Bounds(overrides)


def coerce_bounds(b: Bounds) -> Bounds:
    """Re-validate a Bounds object (mirrors the reference's Bounds.coerce/1).

    ``Bounds`` is a class with a public ``overrides`` mapping, so a caller can hand-craft a
    ``Bounds`` that bypasses ``bounds_new`` — including a WIDENING override (compact_bytes > MAXIMA).
    ``resolve`` trusts the override directly; without this gate the verify/decode paths would honor
    the forged widening (cross-vendor F2). ``coerce_bounds`` re-runs the ``bounds_new`` validation over
    the override map so every entry point that resolves caller-supplied bounds fails closed on a
    forgery.
    """
    from .error import fail

    for key, value in b.overrides.items():
        if not isinstance(value, int) or isinstance(value, bool):
            fail(f"bounds.coerce: non-integer limit {key}")
        if key in FIXED_WIDTH_KEYS:
            if value != MAXIMA[key]:
                fail(f"bounds.coerce: fixed-width key {key} must equal maximum")
        else:
            if value <= 0:
                fail(f"bounds.coerce: non-positive limit {key}")
            if value > MAXIMA[key]:
                fail(f"bounds.coerce: widening limit {key}")
    return b
